### DISCUSSION TITLE
Typescript performance issue illustrated by a modified version of examples/.test/diagnostics-big-router

### DISCUSSION THREAD

**Discussion by thhermansen** [trpc/trpc#5508] [Q&A]
Hi,

We use TRPC in our application and we are experiencing Typescript performance degradation as we have extended the router with queries and mutations. Resolving types for our router now takes 4-8 seconds.

I wanted to test the app TRPC repo uses to demonstrate a big router and found `examples/.test/diagnostics-big-router`.  Without any changes TS resolves the router's type just fine. I am on a M1 Max, with 32 GB of RAM from 2021 and it is close to instant.

However, adding a few more attributes to the router's schema slows it down significantly. With the change in this commit[^1] TS uses 4-6 seconds to come up with routes in `examples/.test/diagnostics-big-router/src/utils/trpc.ts`:

<img width="702" alt="image" src="https://github.com/trpc/trpc/assets/92300/02b982f0-8d8b-4005-ad8c-5a0c34f9ed5f">

Is this what we can expect when it comes to TRPC and performance, or is it something we can do to improve it? We came across https://github.com/algora-io/xtrpc, but that didn't work for us as it failed trying to generate types.

[^1]: Please look at https://github.com/thhermansen/trpc/pull/1 to see the changes I made to the diagnostics-big-router
in order to replicate this.

**Comment by thhermansen** [comment]
@KATT - have you tried this [PR](https://github.com/thhermansen/trpc/pull/1) on your your computer to see how the Typescript performance is?

**Comment by NotLuksus** [comment]
Can confirm that running the changes from the PR degraded my performance aswell.
At my company we are currently running in pretty much the same issue, there are small differences between team members, however for most of us its close to unusable at the moment.
Typing our router as any resolves this issue

**Reply by thhermansen** [reply]
Thank you for checking out my pr and confirming the issue. I have a suspicion we are not alone :(

**Comment by andrewdoro** [comment]
I also have this issue. In my case I am using drizzle and looks like the size of the schema is really slowing the TS.
https://github.com/t3-oss/create-t3-turbo/issues/924

Also disovered this thread on twitter
https://twitter.com/Pruxis/status/1755134044252082488

**Comment by andrewdoro** [comment]
https://github.com/t3-oss/create-t3-turbo/pull/932/files implementing this helped a lot with TS performance
